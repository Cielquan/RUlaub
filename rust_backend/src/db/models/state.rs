use std::fmt::{self, Display, Formatter};

use diesel::prelude::*;
use diesel::result::Error;
use tracing::{debug, instrument, trace};

use super::super::{schema::states, util::last_insert_rowid};

#[derive(Queryable, Debug)]
pub struct State {
    pub id: i32,
    pub state_abbr: String,
    pub state_full: String,
}

impl State {
    pub fn new<'a>(state_abbr: &'a str, state_full: &'a str) -> NewState<'a> {
        NewState {
            state_abbr,
            state_full,
        }
    }
}

impl Display for State {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        if f.alternate() {
            write!(f, "<State {}>", self.state_full)
        } else {
            write!(
                f,
                "<State {} (Abbr: {} | ID: {})>",
                self.state_full, self.state_abbr, self.id
            )
        }
    }
}

#[derive(Insertable, Debug)]
#[table_name = "states"]
pub struct NewState<'a> {
    pub state_abbr: &'a str,
    pub state_full: &'a str,
}

impl Display for NewState<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        if f.alternate() {
            write!(f, "<State {}>", self.state_full)
        } else {
            write!(f, "<State {} (Abbr: {})>", self.state_full, self.state_abbr)
        }
    }
}

impl NewState<'_> {
    #[instrument(skip(self, conn))]
    pub fn save_to_db(self, conn: &SqliteConnection) -> Result<i32, Error> {
        debug!(target: "new_db_entry", "Adding to db: {:?}", &self);
        diesel::insert_into(states::table)
            .values(&self)
            .execute(conn)?;

        trace!(target: "new_db_entry", "Get `last_insert_rowid` for id of: {:#}", &self);
        let id = diesel::select(last_insert_rowid).get_result::<i32>(conn);
        debug!(target: "new_db_entry", "Got ID: <{:?}> for {:#}", id, &self);
        id
    }
}
