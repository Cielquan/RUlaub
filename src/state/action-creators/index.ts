export * from "./configActionCreators";
export * from "./infoPageActionCreators";
export * from "./publicHolidaysDataActionCreators";
export * from "./publicHolidaysDialogActionCreators";
export * from "./schoolHolidaysDataActionCreators";
export * from "./settingsDialogActionCreators";
export * from "./sideMenuActionCreators";
export * from "./vacationDataActionCreators";
