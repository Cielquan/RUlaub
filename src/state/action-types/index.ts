export enum CalendarRowUserMapActionType {
  UPDATE = "update calendar user order",
}

export enum ConfigActionType {
  UPDATE = "update config",
}

export enum InfoPageActionType {
  OPEN = "open InfoPage",
  CLOSE = "close InfoPage",
}

export enum PublicHolidaysDataActionType {
  ADD = "add public holidays data",
  LOAD = "load public holidays data",
  REMOVE = "remove public holidays data",
  UPDATE = "update public holidays data",
}

export enum PublicHolidaysDialogActionType {
  OPEN = "open PublicHolidays",
  CLOSE = "close PublicHolidays",
}

export enum SchoolHolidaysDataActionType {
  ADD = "add school holidays data",
  LOAD = "load school holidays data",
  REMOVE = "remove school holidays data",
  UPDATE = "update school holidays data",
}

export enum SchoolHolidaysDialogActionType {
  OPEN = "open SchoolHolidays",
  CLOSE = "close SchoolHolidays",
}

export enum SettingsDialogActionType {
  OPEN = "open SettingsDialog",
  CLOSE = "close SettingsDialog",
}

export enum SideMenuActionType {
  OPEN = "open SideMenu",
  CLOSE = "close SideMenu",
}

export enum UsersDataActionType {
  ADD = "add users data",
  LOAD = "load users data",
  REMOVE = "remove users data",
  UPDATE = "update users data",
}

export enum UsersDialogActionType {
  OPEN = "open UsersDialog",
  CLOSE = "close UsersDialog",
}

export enum VacationDialogActionType {
  OPEN = "open VacationDialog",
  CLOSE = "close VacationDialog",
}

export enum VacationTypesDataActionType {
  ADD = "add vacation types data",
  LOAD = "load vacation types data",
  REMOVE = "remove vacation types data",
  UPDATE = "update vacation types data",
}

export enum VacationTypesDialogActionType {
  OPEN = "open VacationTypesDialog",
  CLOSE = "close VacationTypesDialog",
}
