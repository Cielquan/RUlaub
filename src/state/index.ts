export * as actionCreators from "./action-creators";
export { default as store } from "./store";
export * from "./reducers";
