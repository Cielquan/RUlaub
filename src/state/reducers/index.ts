import { combineReducers } from "redux";

import dbDataReducer, { initialState as dbDataInitState } from "./dbDataReducer";
import infoPageReducer, { initialState as infoPageInitState } from "./infoPageReducer";
import languageReducer, { initialState as languageInitState } from "./languageReducer";
import localConfigReducer, {
  initialState as localConfigInitState,
} from "./localConfigReducer";
import sideMenuReducer, { initialState as sideMenuInitState } from "./sideMenuReducer";
import themeReducer, { initialState as themeInitState } from "./themeReducer";

const rootReducer = combineReducers({
  dbData: dbDataReducer,
  infoPage: infoPageReducer,
  language: languageReducer,
  localConfig: localConfigReducer,
  sideMenu: sideMenuReducer,
  theme: themeReducer,
});

export type State = ReturnType<typeof rootReducer>;

export const initialState: State = {
  dbData: dbDataInitState,
  infoPage: infoPageInitState,
  language: languageInitState,
  localConfig: localConfigInitState,
  sideMenu: sideMenuInitState,
  theme: themeInitState,
};

export default rootReducer;
