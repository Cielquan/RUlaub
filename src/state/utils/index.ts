interface DBData<T> {
  [k: string]: T;
}

export const addData = <T>(currentData: DBData<T>, dataToAdd: T[]): DBData<T> => {
  const rv: DBData<T> = JSON.parse(JSON.stringify(currentData));
  const highestID = Math.max(...Object.keys(currentData).map((id) => Number(id)));
  dataToAdd.forEach((data, idx) => {
    rv[highestID + idx + 1] = data;
  });
  return rv;
};

export const removeData = <T>(
  currentData: DBData<T>,
  IDsToRemove: string[]
): DBData<T> => {
  const rv: DBData<T> = JSON.parse(JSON.stringify(currentData));
  IDsToRemove.forEach((ID) => {
    delete rv[ID];
  });
  return rv;
};

export const updateData = <T>(
  currentData: DBData<T>,
  updatePayload: [string, T][]
): DBData<T> => {
  const rv: DBData<T> = JSON.parse(JSON.stringify(currentData));
  updatePayload.forEach((update) => {
    const [id, updatedData] = update;
    rv[id] = updatedData;
  });
  return rv;
};
