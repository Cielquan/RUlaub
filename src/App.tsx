import { i18n } from "@lingui/core";
import { I18nProvider } from "@lingui/react";
import { CssBaseline } from "@material-ui/core";
import * as locales from "@material-ui/core/locale";
import { ThemeProvider } from "@material-ui/core/styles";
import React, { ReactElement, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { bindActionCreators } from "redux";

import { actionCreators, State } from "./state";
import useStyles from "./styles";
import createTheme from "./theme";
import { useMountEffect } from "./utils/reactutils";

import Calendar from "./components/Calendar";
import InfoPage from "./components/InfoPage";
import Navbar from "./components/Navbar";
import NewHolidayButton from "./components/NewHolidayButton";
import SideMenu from "./components/SideMenu";

export async function dynamicActivate(locale: string): Promise<void> {
  const { messages } = await import(`./locales/${locale}/messages`);
  i18n.load(locale, messages);
  i18n.activate(locale);
}

function App(): ReactElement {
  const dispatch = useDispatch();
  const { useDarkTheme, useDE, useEN, useLightTheme } = bindActionCreators(
    actionCreators,
    dispatch
  );
  const themeState = useSelector((state: State) => state.theme);
  const langState = useSelector((state: State) => state.language);
  const localConfigState = useSelector((state: State) => state.localConfig);

  const theme = createTheme(themeState, locales[langState.importName]);
  const classes = useStyles();

  useEffect(() => {
    dynamicActivate(langState.locale);
  }, [langState.locale]);

  useMountEffect(() => {
    if (localConfigState.settings.theme !== themeState) {
      (localConfigState.settings.theme === "dark" ? useDarkTheme : useLightTheme)();
    }
    if (localConfigState.settings.language !== langState.locale) {
      (localConfigState.settings.language === "de-DE" ? useDE : useEN)();
    }
  });

  return (
    <I18nProvider i18n={i18n}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        <div data-testid="rulaub-root" className={classes.root}>
          <Navbar />
          <SideMenu />
          <main data-testid="rulaub-main" className={classes.content}>
            <Calendar theme={theme} />
            <NewHolidayButton />
          </main>
        </div>
        <InfoPage />
      </ThemeProvider>
    </I18nProvider>
  );
}

export default App;
