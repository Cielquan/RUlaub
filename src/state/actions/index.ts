import { UserData } from "../../types/usersData.schema";
import {
  CalendarRowUserMapActionType,
  ConfigActionType,
  InfoPageActionType,
  PublicHolidaysDataActionType,
  PublicHolidaysDialogActionType,
  SchoolHolidaysDataActionType,
  SchoolHolidaysDialogActionType,
  SettingsDialogActionType,
  SideMenuActionType,
  UsersDataActionType,
  UsersDialogActionType,
  VacationDialogActionType,
  VacationTypesDialogActionType,
} from "../action-types";
import { ConfigPayload } from "../utils/config";
import {
  PublicHolidaysData,
  PublicHolidayDataPayload,
} from "../utils/publicHolidaysData";
import {
  SchoolHolidaysData,
  SchoolHolidayDataPayload,
} from "../utils/schoolHolidaysData";
import { UsersData, UserDataPayload } from "../utils/usersData";

export type CalendarRowUserMapAction = {
  type: CalendarRowUserMapActionType;
  payload: UsersData;
};

export type ConfigAction = {
  type: ConfigActionType;
  payload: ConfigPayload;
};

export type InfoPageAction = {
  type: InfoPageActionType;
};

export type PublicHolidaysDataLoadAction = {
  type: PublicHolidaysDataActionType.LOAD;
  payload: PublicHolidaysData;
};

export type PublicHolidaysDataUpdateAction = {
  type: PublicHolidaysDataActionType.UPDATE;
  payload: PublicHolidayDataPayload;
};

export type PublicHolidaysDataAction =
  | PublicHolidaysDataLoadAction
  | PublicHolidaysDataUpdateAction;

export type PublicHolidaysDialogAction = {
  type: PublicHolidaysDialogActionType;
};

export type SchoolHolidaysDataLoadAction = {
  type: SchoolHolidaysDataActionType.LOAD;
  payload: SchoolHolidaysData;
};

export type SchoolHolidaysDataUpdateAction = {
  type: SchoolHolidaysDataActionType.UPDATE;
  payload: SchoolHolidayDataPayload;
};

export type SchoolHolidaysDataAction =
  | SchoolHolidaysDataLoadAction
  | SchoolHolidaysDataUpdateAction;

export type SchoolHolidaysDialogAction = {
  type: SchoolHolidaysDialogActionType;
};

export type SettingsDialogAction = {
  type: SettingsDialogActionType;
};

export type SideMenuAction = {
  type: SideMenuActionType;
};

export type UsersDataAddAction = {
  type: UsersDataActionType.ADD;
  payload: UserData[];
};

export type UsersDataLoadAction = {
  type: UsersDataActionType.LOAD;
  payload: UsersData;
};

export type UsersDataRemoveAction = {
  type: UsersDataActionType.REMOVE;
  payload: string[];
};

export type UsersDataUpdateAction = {
  type: UsersDataActionType.UPDATE;
  payload: UserDataPayload[];
};

export type UsersDataAction =
  | UsersDataAddAction
  | UsersDataLoadAction
  | UsersDataRemoveAction
  | UsersDataUpdateAction;

export type UsersDialogAction = {
  type: UsersDialogActionType;
};

export type VacationDialogAction = {
  type: VacationDialogActionType;
};

export type VacationTypesDialogAction = {
  type: VacationTypesDialogActionType;
};
