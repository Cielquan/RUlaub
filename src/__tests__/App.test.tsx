import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

import App from "../App";
import { initialState } from "../state";
import generateMockStore from "../testUtils";

jest.mock("../components/Calendar", () => () => "Calendar");
jest.mock("../components/InfoPage", () => () => "InfoPage");
jest.mock("../components/Navbar", () => () => "Navbar");
jest.mock("../components/NewHolidayButton", () => () => "NewHolidayButton");
jest.mock("../components/SideMenu", () => () => "SideMenu");

describe("<App />", () => {
  const mockStore = generateMockStore(initialState);

  beforeEach(() => {
    mockStore.clearActions();
  });

  it("renders correctly with subComponents mocked as string", () => {
    const tree = renderer
      .create(
        <Provider store={mockStore}>
          <App />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
