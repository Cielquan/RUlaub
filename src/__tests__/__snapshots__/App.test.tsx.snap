// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly with subComponents mocked as string 1`] = `
<div
  className="MuiBox-root css-roynbj"
>
  <div
    className="MuiBox-root css-1jzuftr"
    data-testid="rulaub-root"
  >
    Navbar
    SideMenu
    <main
      className="css-1lnlcp9"
      data-testid="rulaub-main"
    >
      Calendar
      NewHolidayButton
    </main>
  </div>
  InfoPage
  SettingsDialog
</div>
`;
